lib/reader/src/parser.rs

```diff
@@ -42,24 +42,18 @@ pub fn parse_test<'a>(text: &'a str) -> Result<TestFile<'a>> {
     parser.gather_comments(AnyEntity::Function);
 
     let commands = parser.parse_test_commands();
-    let mut isa_spec = parser.parse_isa_specs()?;
-
-    if let isaspec::IsaSpec::Some(ref mut isa_vec) = isa_spec {
-        if isa_vec.len() == 1 {
-            // Remove it from the isa_vec just while we parse functions, avoiding a .clone()
-            parser.unique_isa = Some(isa_vec.remove(0));
-        }
-    }
-
+    let isa_spec = parser.parse_isa_specs()?;
     let preamble_comments = parser.take_comments();
-    let functions = parser.parse_function_list()?;
 
-    // Put unique_isa back in isa_spec if we took it
-    if let isaspec::IsaSpec::Some(ref mut isa_vec) = isa_spec {
-        if let Some(unique_isa) = parser.unique_isa.take() {
-            isa_vec.insert(0, unique_isa);
+    let functions = {
+        let mut unique_isa = None;
+        if let isaspec::IsaSpec::Some(ref isa_vec) = isa_spec {
+            if isa_vec.len() == 1 {
+                unique_isa = Some(&*isa_vec[0]);
```

Comment:
Very cool. You could turn this whole bit into an `IsaSpec::unique_isa()` method.